src/radical/pilot/agent/executing/popen.py

```diff
@@ -703,93 +685,189 @@ def _get_rank_ids(self, n_ranks, launcher):
         else:
             ret += 'test -z "$RP_RANK" && export RP_RANK=0\n'
 
+        # also define a method to sync all ranks on certain events
+        ret += '\nrp_sync_ranks() {\n'
+        ret += '    sig=$1\n'
+        ret += '    echo $RP_RANK >> $sig.sig\n'
+        ret += '    while test $(cat $sig.sig | wc -l) -lt $RP_RANKS; do\n'
+        ret += '        sleep 1\n'
+        ret += '    done\n'
+        ret += '}\n\n'
+
         return ret
 
 
     # --------------------------------------------------------------------------
     #
-    def _get_post_exec(self, task):
+    # rank
+    #
+    def _get_rank_sync(self, sig, ranks):
 
-        ret  = ''
-        cmds = task['description']['post_exec']
-        for cmd in cmds:
-            ret += '%s %s' % (cmd, self._get_check('post_exec'))
+        if ranks == 1:
+            return ''
+
+        # FIXME: make sure that all ranks are alive
+        ret = 'rp_sync_ranks %s\n' % sig.replace(' ', '_')
 
         return ret
 
 
     # --------------------------------------------------------------------------
     #
-    # rank
-    #
-    def _get_pre_rank(self, cmds=None):
+    def _get_env_per_rank(self, task, rank_id, rank, launcher):
```

Comment:
quick comment - I think we don't need this as an env setup, it could be a pre-exec adjustment called at the beginning of `_handle_task` method
```python
cmap = rank['core_map'][0]
if td['cpu_thread_type'] == rpc.OpenMP:
    td['pre_exec'].append('export OMP_NUM_THREADS="%d"\n' % len(cmap))
```
and same for `CUDA_VISIBLE_DEVICES`, but per-rank
```python
if td['gpu_process_type'] == rpc.CUDA:
    # use an abstracted example of "rank_gpu_map" here
    td['pre_exec'].append({str(rank_id): 'export CUDA_VISIBLE_DEVICES=%s' % gpus_per_rank 
                           for rank_id, gpus_per_rank in enumerate(rank_gpu_map)})
```
and then these two env variables will be handled together with other pre-exec commands